File: elixir/sublist/lib/sublist.ex
defmodule Sublist do
  @moduledoc """
  Exercise on list comparison
  """

  @type list_comparison :: :equal | :sublist | :superlist | :unequal
  @doc """
  Returns whether the first list is a sublist or a superlist of the second list
  and if not whether it is equal or unequal to the second list.
  """

  @spec compare(a :: list, b :: list) :: list_comparison
  def compare(list, list), do: :equal

  def compare(a, b) when length(a) < length(b) do
    if sublist?(a, b), do: :sublist, else: :unequal
  end

  def compare(a, b) do
    if sublist?(b, a), do: :superlist, else: :unequal
  end

  defp sublist?(_a, []), do: false
  defp sublist?(a, [_h | t] = b), do: if(List.starts_with?(b, a), do: true, else: sublist?(a, t))

  # defp sublist?([], _b), do: true

  # defp sublist?(a, b) do
  #   Enum.chunk_every(b, length(a), 1, :discard)
  #   |> Enum.any?(&(&1 === a))
  # end
end

File: python/sublist/sublist.py
import typing


T = typing.TypeVar("T")

UNEQUAL, SUBLIST, SUPERLIST, EQUAL = range(4)


def sublist(list_one: list[T], list_two: list[T]) -> typing.Literal[0, 1, 2, 3]:
    """
    Determine whether the two lists are equal, the first is a sublist
    or a superlist of the second, or they are unequal.
    """
    if list_one == list_two:
        return EQUAL
    if contains(list_two, list_one):
        return SUPERLIST
    if contains(list_one, list_two):
        return SUBLIST
    return UNEQUAL


def contains(list_one: list[T], list_two: list[T]) -> bool:
    """
    List A is a sublist of B if B contains a contiguous sub-sequence of values equal to A.
    """
    list_one_len = len(list_one)
    for start in range(len(list_two) - list_one_len + 1):
        if list_two[start : start + list_one_len] == list_one:
            return True
    return False
